cpp/src/arrow/util/tracing_internal.h

```diff
@@ -97,6 +98,57 @@ AsyncGenerator<T> WrapAsyncGenerator(AsyncGenerator<T> wrapped,
     return fut;
   };
 }
+
+/// \brief Start a new span for each invocation of a generator.
+///
+/// The parent span of the new span will be the currently active span
+/// (if any) as of when WrapAsyncGenerator was itself called.
+template <typename T>
+AsyncGenerator<T> WrapAsyncGenerator(AsyncGenerator<T> wrapped,
+                                     const std::string& span_name) {
+  opentelemetry::trace::StartSpanOptions options;
+  options.parent = GetTracer()->GetCurrentSpan()->GetContext();
+  return WrapAsyncGenerator(std::move(wrapped), std::move(options), span_name);
+}
+
+/// \brief End the given span when the given async generator ends.
+///
+/// The span will be made the active span each time the generator is called.
+template <typename T>
+AsyncGenerator<T> TieSpanToAsyncGenerator(
+    AsyncGenerator<T> wrapped,
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span) {
+  return [=]() mutable -> Future<T> {
+    auto scope = GetTracer()->WithActiveSpan(span);
+    auto fut = wrapped();
+    fut.AddCallback([span](const Result<T>& result) {
+      if (!result.ok() || IsIterationEnd(*result)) {
+        MarkSpan(result.status(), span.get());
+        span->End();
+      }
+    });
+    return fut;
```

Comment:
Maybe use `Then`?  The timing can't really be guaranteed otherwise.  Consuming callbacks could run before this callback does.  But I suppose that isn't too critical.  I don't have enough headspace around spans yet to know if we need to worry about the timing (e.g. will the consumer start a new active span and then this one ends that the wrong one?)

Reply:
Spans are only ended if `End` is called or if all references are freed. In particular `span` is a `shared_ptr`. The "active span" is only relevant for automatic span linkage. So starting a new active span does not necessarily end the old span. We could certainly use `Then` here, but I don't think there's any risk of ending the wrong span on accident.

Reply:
Also, this callback is added before any consumer has a chance to add callbacks right? Or is the order of callbacks undetermined?

Reply:
The order of callback can be undetermined in the event of:

```
Future<> fut = DoThing(); // Future unfinished at this point
fut.AddCallback(...); // Callback actually added
// Future finishes but callbacks haven't run yet...
fut.AddCallback(...); // Runs immediately
```

This has caused several subtle bugs and left me paranoid of `AddCallback`.  That being said, if the ordering truly doesn't matter, then it should be fine here.

Reply:
Ah, right. We can make this `Then` just to be sure, then, I don't think it makes a big difference.